Make the `EventSource` generator incremental et.al.

teo-tsirpanis opened the pull request:
The `EventSource` generator was made incremental and its output is marked as auto-generated. (TIL GitHub doesn't let me reopen a PR if I have force-pushed its branch)

Fixes #64563.

c.c. @danmoseley

teo-tsirpanis committed:
Mark the EventSource generator's output as auto-generated.

Fixes #64563.

teo-tsirpanis committed:
Make the EventSource generator incremental.

stephentoub commented on src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.Emitter.cs:
```diff
@@ -12,76 +12,72 @@ namespace Generators
 
-                    _builder.AppendLine("using System;");
-                    GenType(ec);
+                sb.AppendLine(s_header);
```
Why put s_header in a static rather than just having this be: `sb.AppendLine("// <auto-generated/>").AppendLine();`?

teo-tsirpanis replied:
I followed what other source generators with more complex headers were doing, but I made it simpler.

stephentoub commented on src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.Parser.cs:
```diff
@@ -16,58 +17,47 @@ namespace Generators
-            }
-
-            public EventSourceClass[] GetEventSourceClasses(List<ClassDeclarationSyntax> classDeclarations)
+            public static ImmutableArray<EventSourceClass> GetEventSourceClasses(ImmutableArray<ClassDeclarationSyntax> classDeclarations, Compilation compilation, CancellationToken ct)
```
ct => cancellationToken

teo-tsirpanis replied:
Done.

stephentoub commented on src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.cs:
```diff
@@ -32,69 +32,53 @@ public partial class EventSourceGenerator : ISourceGenerator
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, _) => GetSemanticTargetForGeneration(x))
+                .Where(x => x is not null)
+                .Collect()
+                .Combine(context.CompilationProvider)
```
@chsienki can correct me if I'm wrong, but I believe you'll need to do something like:
https://github.com/dotnet/runtime/blob/6c50d9fad7217b0eeed2b27ca6ab9ab34bc0dd1f/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs#L40-L45
to avoid causing regeneration to occur on _any_ change to the compilation.

teo-tsirpanis replied:
It was not as simple since we had to compare an immutable array of class declerations, but done.

teo-tsirpanis commented on src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.Parser.cs:
```diff
@@ -16,58 +17,47 @@ namespace Generators
                         {
                             // be nice and stop if we're asked to
-                            return results?.ToArray() ?? Array.Empty<EventSourceClass>();
+                            return results?.ToImmutable() ?? ImmutableArray<EventSourceClass>.Empty;
```
And once @chsienki comes, couldn't we return an empty array here since we were asked to cancel?

teo-tsirpanis replied:
Never mind, let's stick with the status quo.

teo-tsirpanis committed:
Address PR feedback.

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
-                return;
-            }
+            // Only add classes annotated [EventSourceAutoGenerate] to reduce busy work.
+            const string EventSourceAttribute = "EventSourceAutoGenerateAttribute";
```
Note that this wont work in the presence of type aliases. That might be an acceptable tradeoff, but something to be aware of.

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
+        private static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
+            node is ClassDeclarationSyntax x && x.AttributeLists.Count > 0;
+
+        private static ClassDeclarationSyntax? GetSemanticTargetForGeneration(GeneratorSyntaxContext context)
```
This doesn't appear to be actually doing any semantic work? It's just checking the syntax. In which case you should move the logic into the syntactic check.

Syntax checks are only re-performed on files that are modified, whereas semantic checks have to run every time any edit is made (as you might be looking something up in another file). Moving this work to the syntactic check will give you significantly improved perf in the IDE on bigger projects.

teo-tsirpanis replied:
So I should move `GetSemanticTargetForGeneration` into `IsSyntaxTargetForGeneration`? [I read that](https://andrewlock.net/creating-a-source-generator-part-1-creating-an-incremental-source-generator/) `CreateSyntaxProvider`'s predicate must be very fast, but it looks like that the litmus test is whether we use the context's `SemanticModel`. Thanks, I will fix it.

teo-tsirpanis replied:
(or even better I will actually use the semantic model and correctly detect the attribute)

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:
```diff
@@ -16,58 +17,47 @@ namespace Generators
+                ImmutableArray<EventSourceClass>.Builder? results = null;
                 // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
-                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax>? group in classDeclarations.GroupBy(x => x.SyntaxTree))
+                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax> group in classDeclarations.GroupBy(static x => x.SyntaxTree))
```
Ideally, you would move this work into the semantic check part of the SyntaxProvider. That way you don't need to new up a semantic model (there already is one) which allows it to be shared among other generators.

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:
```diff
@@ -16,58 +17,47 @@ namespace Generators
                         {
                             // be nice and stop if we're asked to
-                            return results?.ToArray() ?? Array.Empty<EventSourceClass>();
+                            return results?.ToImmutable() ?? ImmutableArray<EventSourceClass>.Empty;
```
```suggestion
                            cancellationToken.ThrowIfCancellationRequested();
```

The generator driver will handle cancellation exceptions for you and keep the cached values. Returning empty results counts as a 'success' run, and the driver will cache the empty results, meaning you need to do more work next time around.

stephentoub commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
+                .Collect()
+                .Combine(context.CompilationProvider)
+                // Use a custom comparer that ignores the compilation.
+                .WithComparer(ClassDeclarationsEqualityComparer.Instance)
```
@chsienki has encouraged us in the past to do this, and we do in our other generators. Why is this case different?

sharwell replied:
This case is not different. The RegexGenerator is also broken (e.g. if you start with a project that allows unsafe, and then you alter the project where the _only change_ is to disable unsafe code, RegexGenerator will fail to detect the change and will keep the original code).

stephentoub replied:
IIRC, @chsienki recommended doing this because otherwise any change anywhere in the project (or maybe file) would invalidate caching, and thus basically caching was useless.  Is that not the case?

sharwell replied:
The correct approach for RegexGenerator is to extract a subset of information from the compilation into a comparable, self-contained data model (e.g. one property it would have is `AllowsUnsafe`). Here's an example from another generator:

https://github.com/microsoft/vs-extension-testing/blob/6e2208049bcc1581883cd7d146a2f37e490b9fb0/src/Microsoft.VisualStudio.Extensibility.Testing.SourceGenerator/TestServicesSourceGenerator.cs#L617-L632

In practice, this data extract step will run fairly often, but it will generally produce the same result as the previous incremental generation and thus not trigger downstream steps to run.

teo-tsirpanis replied:
Thanks @sharwell, I will keep that in mind.

And I am interested to fix the other generators accordingly.

teo-tsirpanis committed:
Move the semantic checks and data gathering earlier in the pipeline.

The code was significantly shrunk.

teo-tsirpanis committed:
Stop nesting classes inside the generator.

Not actually needed, and reduces indentation.

teo-tsirpanis commented:
I updated the generator to only use the semantic model provided by `CreateSyntaxProvider`'s transform delegate, completely removing the need for injecting the `Compilation` into the pipeline, and drastically simplifying it.

Can you review it again?

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,69 +33,14 @@ public partial class EventSourceGenerator : ISourceGenerator
+            context.RegisterSourceOutput(eventSourceClasses, static (spc, x) => EmitSourceFile(spc, x));
         }
 
         private sealed class EventSourceClass
```
Make this `IEquatable<>` or supply a custom comparer, so code won't get regenerated when you end up with the same set of information.

sharwell replied:
💡 Ideally, make it a record.

teo-tsirpanis replied:
Done.

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,69 +33,14 @@ public partial class EventSourceGenerator : ISourceGenerator
-                    }
-                }
-            }
+            context.RegisterSourceOutput(eventSourceClasses, static (spc, x) => EmitSourceFile(spc, x));
```
Nit:

```suggestion
            context.RegisterSourceOutput(eventSourceClasses, EmitSourceFile);
```

To avoid the extra method generation.

sharwell replied:
Is the repository already building with the newer version of the compiler that caches this delegate?

teo-tsirpanis replied:
I forgot that Roslyn started caching the delegate on method groups, I converted the lambdas (even `IsSyntaxTargetForGeneration` where I gave it an unused cancellation token). Even uncached it won't hurt much performance-wise, it's just some delegates in the `Initialize` method.

chsienki commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs:
```diff
@@ -32,69 +33,14 @@ public partial class EventSourceGenerator : ISourceGenerator
-                            }
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, cancellationToken) => GetSemanticTargetForGeneration(x, cancellationToken))
```
Nit:
```suggestion
                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), GetSemanticTargetForGeneration)
```

chsienki reviewed (approved):
You'll want to implement `IEquatable<>` on the EventSourceClass to ensure you don't do duplicate work unnecessarily, but otherwise LGTM.

sharwell commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Emitter.cs:
```diff
@@ -12,97 +12,86 @@ namespace Generators
 
-                    _builder.AppendLine("using System;");
-                    GenType(ec);
+        private static void EmitSourceFile(SourceProductionContext spc, EventSourceClass ec)
```
💡 Convention for analyzers/code fixes/refactorings/source generators is to use `context` as the parameter name for all input contexts.

```suggestion
        private static void EmitSourceFile(SourceProductionContext context, EventSourceClass ec)
```

teo-tsirpanis replied:
Done.

sharwell commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:
```diff
@@ -16,180 +17,140 @@ namespace Generators
     {
-        private sealed class Parser
+        private static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
+            node is ClassDeclarationSyntax x && x.AttributeLists.Count > 0;
```
```suggestion
            node is ClassDeclarationSyntax { AttributeLists.Count: > 0 };
```

teo-tsirpanis replied:
Done.

sharwell reviewed (commented):
This approach is much better! So glad to see the compilation issue resolved 👍

teo-tsirpanis committed:
Address remaining PR feedback.

Make EventSourceClass a record, use method groups instead of lambdas, and rename a parameter.

teo-tsirpanis commented:
All PR feedback was addressed.

For the record, source generators (especially the incremental ones) are awesome.

teo-tsirpanis committed:
Remove a couple of unused files from the project.

tommcdon commented:
@stephentoub @noahfalk @hoyosjs

tommcdon commented:
@stephentoub @noahfalk @hoyosjs just a friendly ping, any concerns with moving forward with this change?

stephentoub commented on src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:
```diff
@@ -3,193 +3,147 @@
+                        continue;
+                    }
 
-                INamedTypeSymbol? eventSourceAttribute = _compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");
```
Why are we removing the use of GetBestTypeByMetadataName?  Is this different from the other conversation on the regex generator, where we opted to keep its usage?

teo-tsirpanis replied:
This PR was opened before that conversation. I will revert back to `GetBestTypeByMetadataName` soon.

stephentoub replied:
Ok, thanks.

noahfalk reviewed (approved):
I only skimmed, but LGTM. Thanks!

danmoseley commented:
Is this good to merge now?

chsienki commented:
@danmoseley good from the compiler side.

chsienki commented:
@danmoseley Looks good from the compiler side, modulo the revert of `GetBestTypeByMetadataName`. Not a blocker, but as discussed on the regex generator PR it can be a subtle breaking change so suggest switching it back.

teo-tsirpanis commented:
OK, I reverted to `GetBestTypeByMetadataName`.

danmoseley commented:
@noahfalk @hoyosjs is this mergeable (after build break is fixed..)

teo-tsirpanis commented:
Build what? Oh crap, I had built it from VS and it worked. I will take a look soon. 😥

teo-tsirpanis committed:
Fix IsExternalInit's path.

teo-tsirpanis commented:
Turns out #66434 changed the location of `IsExternalInit.cs`. It builds on my machine after a rebase. Let's see now...

teo-tsirpanis commented:
Ahhh, what went wrong this time?

@chsienki I removed `GetBestTypeByMetadataName`. This generator is only internally used by corelib, so there is no need for keeping backwards compatibility over an edge-case scenario.

noahfalk commented:
> @noahfalk @hoyosjs is this mergeable (after build break is fixed..)

Yep!

danmoseley commented:
Thanks @teo-tsirpanis !

teo-tsirpanis commented:
BTW are you still on vacation @noahfalk? Perhaps you forgot to clear your GitHub profile status.

noahfalk commented:
nope I've been back a good while, thanks for the heads up : )
